Reporting build status while no permission

shreyaspapi opened the pull request:
In Gitlab while working with upstream-project we don't have access to the merge request commit coming from fork repos. To set commit status we need access to the repository. 

So, if the reporting fails we can report the state using the MR comment and request access to the forked repository ([docs](https://docs.gitlab.com/ee/api/access_requests.html)). 

- [x] Adding the functionality to report build status using MR comment.
- [x] Fixed enqueue error #741 
- [x] Requesting for project access (with right permissions).

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +389,28 @@ def _report(
+            self.status_reporter.report(
+                description=description, state=state, url=url, check_names=check_names,
+            )
+        except Exception as e:
```
Catching such general exception can be dangerous -- there are a lot of other problems we can get.

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +389,28 @@ def _report(
+        except Exception as e:
+            logger.info("Unable to report status ", e)
+            self.report_status_by_comment(
+                int(self.metadata.identifier), state, url, check_names
```
- Please, use key-word arguments for better readability.
- `identifier` can be non-numeric for non-merge-request events.

shreyaspapi replied:
I could only find identifier as common between all events.

lachmanfrantisek replied:
My point was about `int(self.metadata.identifier)` that we can get non-numeric value from `self.metadata.identifier`. What will cause an exception...

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +389,28 @@ def _report(
+                int(self.metadata.identifier), state, url, check_names
+            )
+
+    # Reporting build status if no permission to the fork project
```
Use docstring for that information. (The `"""something"""` string bellow the function/class header.)

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +95,35 @@ def local_project(self) -> LocalProject:
+    def request_project_access(self) -> None:
+        try:
+            self.base_project.request_access()
+        except Exception:
```
I would use the most specific exception possible and ideally check also content. We can get more exception here. (E.g. network/authentication problems...).

shreyaspapi replied:
Yes added more specific exceptions.

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -207,9 +240,14 @@ def job_tests(self) -> Optional[JobConfig]:
-            self._status_reporter = StatusReporter(
-                self.project, self.metadata.commit_sha, self.metadata.pr_id
-            )
+            if isinstance(self.project, GitlabProject):
+                self._status_reporter = StatusReporter(
+                    self.base_project, self.metadata.commit_sha, self.metadata.pr_id
+                )
+            else:
+                self._status_reporter = StatusReporter(
+                    self.project, self.metadata.commit_sha, self.metadata.pr_id
+                )
```
Maybe this to show the only difference more clearly?

```suggestion
            self._status_reporter = StatusReporter(
                project=self.base_project if isinstance(self.project, GitlabProject) else self.project,
                commit_sha=self.metadata.commit_sha,
                pr_id=self.metadata.pr_id
            )
```

shreyaspapi replied:
Done!

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +411,59 @@ def _report(
-        )
+        if (
+            isinstance(self.project, GitlabProject)
+            and not self.check_project_permission
```
The name `check_project_permission` looks more like a method. With properties, I would use some `is_*` form of the name. E.g. `is_reporting_allowed`, or `is_fork_status_allowed`,...

shreyaspapi replied:
Yup, is_reporting_allowed is better name here. Changed it!

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +411,59 @@ def _report(
+        """
+        comment_msg = f"[{check_names}]({url}): {state.name.upper()}"
+
+        all_pr_comments = self.project.get_pr(pr_id=self.pr_id)._get_all_comments()[
+            ::-1
+        ]
+
```
Don't we have some argument to do this with the public method?

shreyaspapi replied:
Thanks! Found it.

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +411,59 @@ def _report(
+            ::-1
+        ]
+
+        if isinstance(check_names, list):
+            return
```
Can you please document this a bit?

shreyaspapi replied:
Found the method get_comments(), changed the comment reporting a bit. Also added few comments.

lachmanfrantisek commented on packit_service/worker/reporting.py:
```diff
@@ -95,9 +96,16 @@ def set_status(
+        except gitlab.exceptions.GitlabHttpError:
+            # Ignoring Gitlab 'enqueue' error
+            # https://github.com/packit-service/packit-service/issues/741
+            pass
```
What is done when we get that exception? Nothing is changed on GitLab side?

What can we do instead to show that info to the user? What about having a unique status for each step of our workflow? e.g. `packit-service/copr-build-srpm`, `packit-service/copr-build-fedora-31-rpm` (We do that in the past..;)

lachmanfrantisek replied:
Do we want to solve this here or make a separate PR for this. We have multiple possibilities for that so it's worth a separate discussion.

shreyaspapi replied:
I will make a different pr for this.

lachmanfrantisek replied:
Thanks!

shreyaspapi commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +96,39 @@ def local_project(self) -> LocalProject:
+    def base_project(self) -> GitProject:
+        """
+        Getting the source project info from PR,
+        In case of build events we loose the source info.
```
> * If you mean results events, we can get that info from the database.

Yeah, I meant the result event. Okay, I will try to use the info from the database.

shreyaspapi replied:
We don't store base_project info in database.

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +96,35 @@ def local_project(self) -> LocalProject:
+        In case of build events we loose the source info.
+        """
+        if self._base_project is None:
+            self._base_project = self.project.get_pr(pr_id=self.pr_id).source_project
```
This will not work for non-pr cases. We should return `base_project` in that case.

shreyaspapi replied:
The best I could think of was by cathing the value error here which will come from pr_id, in case of non-pr events.

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +96,41 @@ def local_project(self) -> LocalProject:
+        In case of build events we loose the source info.
+        """
+        if self._base_project is None:
+            try:
+                self._base_project = self.project.get_pr(
+                    pr_id=self.pr_id
+                ).source_project
+            # pr_id will result in value error for non-pr events
+            except (ValueError, gitlab.exceptions.GitlabGetError):
+                return self.project
```
What about this?

```suggestion
            if self.pr_id:
                self._base_project = self.project.get_pr(
                    pr_id=self.pr_id
                ).source_project
            else:
                self._base_project = self.project
```

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +96,43 @@ def local_project(self) -> LocalProject:
+    def pr_id(self) -> int:
+        if self._pr_id is None:
+            try:
+                self._pr_id = int(self.metadata.event_dict["identifier"])
```
@lbarcziova do we have `pr_id` in the `event_dict`? (So we don't need to do this `try-except`.) Thanks

lbarcziova replied:
it depends on the content of the original Event, but in EventData (self.metadata) we are trying to get both [pr_id](https://github.com/packit-service/packit-service/blob/398d0d20060ea3af92d21d9dab9b3adef599f365/packit_service/service/events.py#L186)  and [identifier](https://github.com/packit-service/packit-service/blob/398d0d20060ea3af92d21d9dab9b3adef599f365/packit_service/service/events.py#L188), so this could be definitely simplified

lachmanfrantisek replied:
@lbarcziova Thanks.

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +417,30 @@ def _report(
+        """
+        build_states = [CommitStatus.error, CommitStatus.success, CommitStatus.failure]
+
+        if isinstance(check_names, str):
```
What about the `check_names is None` scenario. (At least from the type annotation, it looks like we can get that.)

shreyaspapi replied:
Ah, I have changed the type annotation.

shreyaspapi replied:
Solved it here - [R438-R439](https://github.com/packit/packit-service/pull/740/files#diff-599fa650469aeda8bea187fa74ed9502R438-R439)

lachmanfrantisek replied:
If I read that correctly, we don't want to process status that is final (`error`/`success`/`failure`).

- What will be done for those "final" states?
- Is it possible to move this logic (= `if state not in ...` condition) out of this method? Is it too complicated?
- Can we add a comment and/or update docstring to say why we are doing this?

shreyaspapi replied:
@lachmanfrantisek 
We only want to report status by comment if it's a final state (error / success / failure).

```
if state not in build_states or check_names is None:
    return
```
The above condition was still valid, as we are not doing anything if `state` has any value other then (error/success/failure).

Anyways I have updated and moved that part as suggested. 

I think this might be easier to read -
```
if state in build_states and check_names is not None:
    self.status_reporter.report_status_by_comment(state, url, check_names)
```

lachmanfrantisek replied:
Ah, yes. My bad...;)

Thanks!

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -373,9 +417,30 @@ def _report(
+        if isinstance(check_names, str):
+            check_names = [check_names]
+
+        for check in check_names:
+            comment_msg = f"[{check}]({url}): {state.name.upper()}"
+            if state in build_states:
+                self.project.pr_comment(self.pr_id, comment_msg)
```
Can we merge those to one comment to not be so spammy?

shreyaspapi replied:
Done!

shreyaspapi commented:
Thanks for the review! @lachmanfrantisek 
I have addressed all the comments here.

>What can we do instead to show that info to the user? What about having a unique status for each step of our workflow? e.g. >packit-service/copr-build-srpm, packit-service/copr-build-fedora-31-rpm (We do that in the past..;)

The only thing which is left here is to send unique build status for each step of the workflow, I will have different for it!

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -375,9 +418,34 @@ def _report(
-        self.status_reporter.report(
-            description=description, state=state, url=url, check_names=check_names,
-        )
+        if isinstance(self.project, GitlabProject) and not self.is_reporting_allowed:
+            # "Packit-User" does not have access to project.
+            self.request_project_access()
+            self.report_status_by_comment(state, url, check_names)
+        else:
+            self.status_reporter.report(
+                description=description, state=state, url=url, check_names=check_names,
+            )
```
What about moving the `report_status_by_comment` to `status_reporter` class?

shreyaspapi replied:
Yes, I have moved `report_status_by_comment ` to `StatusReporter` class.

lachmanfrantisek reviewed (commented):
Looks good!

I have only a few code-style suggestions.

lachmanfrantisek commented on packit_service/service/api/webhooks.py:
```diff
@@ -219,25 +219,25 @@ def create_confidential_issue_with_token(self):
                 title="Packit-Service Authentication",
                 body=f"To configure Packit-Service with `{repo_name}` you need to\n"
-                f"configure the webhook settings. Head to {http_url}/hooks and add\n"
+                f"configure the webhook settings. Head to {http_url[:-4]}/hooks and add\n"
```
You can use [project.get_web_url method](https://github.com/packit/ogr/blob/176add79eeb7d71e765550da76c9cdc8aced9e92/ogr/services/gitlab/project.py#L554) instead for better readability:

```suggestion
                f"configure the webhook settings. Head to {project.get_web_url()}/hooks and add\n"
```

lachmanfrantisek commented on packit_service/service/api/webhooks.py:
```diff
@@ -219,25 +219,25 @@ def create_confidential_issue_with_token(self):
                 f"[{packit_user}](https://gitlab.com/{packit_user}) `admin` permissions\n"
                 f"on the {namespace}/{repo_name} project. You can add the rights by clicking\n"
-                f"[here]({http_url}/-/project_members).",
+                f"[here]({http_url[:-4]}/-/project_members).",
```
```suggestion
                f"[here]({project.get_web_url()}/-/project_members).",
```

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +97,47 @@ def local_project(self) -> LocalProject:
             )
         return self._local_project
 
+    @property
+    def is_gitlab_instance(self) -> bool:
+        if self._is_gitlab_instance is None:
+            self._is_gitlab_instance = isinstance(self.project, GitlabProject)
+
+        return self._is_gitlab_instance
```
I am not sure if this helps too much. I am leaving that to your judgement if it is clearer now. (=You don't need to change that if you want..;)

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -91,6 +97,47 @@ def local_project(self) -> LocalProject:
+    def request_project_access(self) -> None:
+        try:
+            self.base_project.request_access()
+        except GitlabAPIException:
+            logger.info("Access already requested")
```
Is there something in the exception instance or exception string representation what can we use to filter only those that mean `Access already requested`?

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -377,12 +428,27 @@ def _report(
+            # "Packit-User" does not have access to project.
+
+            self.request_project_access()
+            build_states = [
+                CommitStatus.error,
+                CommitStatus.success,
+                CommitStatus.failure,
+            ]
+
+            # We are only commenting final states to avoid multiple comments for a build
+            # Ignoring all other states eg. pending, running
+            if state in build_states and check_names is not None:
+                self.status_reporter.report_status_by_comment(state, url, check_names)
```
```suggestion
            final_commit_states = [
                CommitStatus.error,
                CommitStatus.success,
                CommitStatus.failure,
            ]

            # We are only commenting final states to avoid multiple comments for a build
            # Ignoring all other states eg. pending, running
            if state in final_commit_states and check_names is not None:
                self.status_reporter.report_status_by_comment(state, url, check_names)
```

lachmanfrantisek commented on packit_service/worker/reporting.py:
```diff
@@ -87,6 +88,25 @@ def __set_pull_request_status(
+        if isinstance(check_names, str):
+            check_names = [check_names]
+
+        comment_msg = [
+            f"[{check}]({url}): {state.name.upper()}" for check in check_names
+        ]
```
We can use a table here (not need to be done here). The table looks better but only in case we don't need to provide any logs inside.

shreyaspapi replied:
Yup we don't, I have changed it!

lachmanfrantisek commented on packit_service/worker/build/build_helper.py:
```diff
@@ -377,12 +428,27 @@ def _report(
-            url=url,
-            check_names=check_names,
-        )
+        if self.is_gitlab_instance and not self.is_reporting_allowed:
```
What about switching the order of those two `if` branches to have the shorter one in the beginning?

So, instead of this:
```python
if something:
    # long
else:
    # short
```

use this:

```python
if not something:
    # short
    return

# long
```

shreyaspapi replied:
👍

lachmanfrantisek reviewed (commented):
Nice job!

I have a few notes but mostly styling nits or suggestions for the future. No worry about the number of comments..;)

lachmanfrantisek commented on files/docker/Dockerfile.tests:
```diff
@@ -1,6 +1,6 @@
 # For running tests locally, see check_in_container target in Makefile
 
-FROM docker.io/usercont/packit-service-worker:dev
+FROM docker.io/usercont/packit-service-worker:stg
```
Can you revert that, please?

lachmanfrantisek commented on packit_service/worker/reporting.py:
```diff
@@ -87,6 +88,24 @@ def __set_pull_request_status(
+
+        if isinstance(check_names, str):
+            check_names = [check_names]
+        comment_msg = [
+            f"| [{check}]({url}) | {state.name.upper()} |" for check in check_names
+        ]
```
We are probably missing the header of the table. It should look like this:

```
| some text | in header |
| ------------- | ------------ |
| one | line |
| another | line |
```
| some text | in header |
| ------------- | ------------ |
| one | line |
| another | line |

or empty header:

```
| | |
| ------------- | ------------ |
| one | line |
| another | line |
```
| | |
| ------------- | ------------ |
| one | line |
| another | line |

Without that it looks like this:
```
| one | line |
| another | line |
```
| one | line |
| another | line |

lachmanfrantisek commented on packit_service/worker/reporting.py:
```diff
@@ -87,6 +88,26 @@ def __set_pull_request_status(
+            f"| [{check}]({url}) | {state.name.upper()} |" for check in check_names
+        ]
+
+        self.project.pr_comment(self.pr_id, "\n\n".join(header + comment_msg))
```
I think we need only one new line here:
```suggestion
        self.project.pr_comment(self.pr_id, "\n".join(header + comment_msg))
```

lachmanfrantisek reviewed (approved):
Good job! Thank you!

shreyaspapi commented:
@lachmanfrantisek problems with tests! We need to wait.

TomasTomecek commented:
Wow, merged \o/ good job Shreyas!

![\o/](https://media.giphy.com/media/l2QDUxcli4pw5yiOs/giphy.gif)
